src/week2.hs

```diff
@@ -0,0 +1,168 @@
+exactMatches :: Code -> Code -> Int
+
+exactMatches [] _ = 0
+exactMatches _ [] = 0
+exactMatches (x:xs) (x':xs') 
+    |x == x' = 1 + (exactMatches xs xs')
+    |otherwise = (exactMatches xs xs')
+
+-- Exercise 2 -----------------------------------------
+-- Write a function that returns the number of total matches between the secret code and the guess
+
+-- helper
+sumLists :: [Int] -> [Int] -> [Int]
+sumLists [] _ = []
+sumLists _ [] = []
+sumLists (x:xs) (x':xs') = (x + x'):(sumLists xs xs')
+
+countColors :: Code -> [Int]
+
+countColors [] = [0,0,0,0,0,0]
+countColors (x:xs)
+    | x == Red    = sumLists [1,0,0,0,0,0] (countColors xs)
+    | x == Green  = sumLists [0,1,0,0,0,0] (countColors xs)
+    | x == Blue   = sumLists [0,0,1,0,0,0] (countColors xs)
+    | x == Yellow = sumLists [0,0,0,1,0,0] (countColors xs)
+    | x == Orange = sumLists [0,0,0,0,1,0] (countColors xs)
+    | otherwise   = sumLists [0,0,0,0,0,1] (countColors xs)
+
+sumMin :: [Int] -> [Int] -> Int
+sumMin [] _ = 0
+sumMin _ [] = 0
+sumMin (x:xs) (x':xs') = (minimum [x, x']) + (sumMin xs xs')
+
+matches :: Code -> Code -> Int
+matches x y = sumMin (countColors x) (countColors y) 
+
+-- Exercise 3 -----------------------------------------
+-- Construct a Move from a guess given the actual code
+getMove :: Code -> Code -> Move
+getMove s g = Move (g) (exactMatches s g)  ((matches s g)  - (exactMatches s g))
+
+-- Or:
+getMove1 :: Code -> Code -> Move
+getMove1 s g = Move g e (m-e) 
+        where e = exactMatches s g
+              m = matches s g
+
+-- Exercise 4 -----------------------------------------
+-- A code is consistent with a Move if the Code could have been the secret that generated that move
+-- Write the function isConsistent
+isConsistent :: Move -> Code -> Bool
+
+isConsistent (Move (c) (x) (y)) c' 
+    |(exactMatches c c') == x && ((matches c c') - (exactMatches c c')) == y = True
+    |otherwise                                                               = False
+    
+-- Or:
+isConsistent1 :: Move -> Code -> Bool
+isConsistent1 (Move (c) (x) (y)) c'  = e == x && (m - e) == y
+    where e = exactMatches c c'
+          m = matches c c'
+
+-- Exercise 5 -----------------------------------------
+-- Write a function that filters a list of Codes to only contain those that are consistent with a given Move
+filterCodes :: Move -> [Code] -> [Code]
+filterCodes _ [] = []
+filterCodes m (c:cs) = if (isConsistent m c) then (c : filterCodes m cs)
+                                             else filterCodes m cs
+-- or, more concisely:
+filterCodes1 :: Move -> [Code] -> [Code]
+filterCodes1 m c = filter (isConsistent m) c
+
+-- Exercise 6 -----------------------------------------
+-- Write a function that outputs all possible code permutations of a given length
+
+pegs :: Code
+pegs = [Red, Green, Blue, Yellow, Orange, Purple]
+baseCodes :: [Code]
+baseCodes = [[Red], [Green], [Blue], [Yellow], [Orange], [Purple]]
```

Comment:
Good thinking. I think you can make this even more concise, if you start with zero-length permutation as your base case.

Reply:
Thanks for the comments! I've made some changes accordingly